Jenkinsfile

```diff
@@ -7,95 +7,111 @@ def propagateParamsToEnv() {
-             name: 'DESTROY_ON_FAILURE'),
-      choice(choices: ["true", "false"].join("\n"),
-             description: 'Abort all tests upon first failure.',
-             name: 'FAIL_FAST'),
-    ]),
+      disableConcurrentBuilds(),
+      parameters([
+        // WARNING: changing parameters will not affect the next build, only the following one
+        // see issue #1315 or https://stackoverflow.com/questions/46680573/ -- 2020-04 walt
+        choice(choices: ["run", "skip"].join("\n"),
+          // defaultValue is not applicable to choices. The first choice will be the default.
+          description: 'Run or skip robotest system wide tests.',
+          name: 'RUN_ROBOTEST'),
+        choice(choices: ["true", "false"].join("\n"),
+          description: 'Destroy all VMs on success.',
+          name: 'DESTROY_ON_SUCCESS'),
+        choice(choices: ["true", "false"].join("\n"),
+          description: 'Destroy all VMs on failure.',
+          name: 'DESTROY_ON_FAILURE'),
+        choice(choices: ["true", "false"].join("\n"),
+          description: 'Abort all tests upon first failure.',
+          name: 'FAIL_FAST'),
+      ]),
   ])
 }
 
-timestamps {
-  node {
-    stage('checkout') {
-      checkout scm
-      sh "git submodule update --init --recursive"
-      sh "sudo git clean -ffdx" // supply -f flag twice to force-remove untracked dirs with .git subdirs (e.g. submodules)
-    }
-    stage('params') {
-      // For try builds, we DO NOT want to overwrite parameters, as try builds
-      // offer a superset of PR/nightly parameters, and the extra ones will be
-      // lost when setRobotestParameters() is called -- 2020-04 walt
-      echo "Jenkins Job Parameters:"
+// robotest() defines the robotest pipeline.  It is expected to be run after
+// the gravity build, as it implicitly relies on artifacts & state from those targets.
+//
+// Per https://plugins.jenkins.io/pipeline-stage-view/: If we want to visualize
+// dynamically changing stages, it is better to make it conditional to execute the
+// stage contents, not conditional to include the stage.
+def robotest() {
+  stage('set robotest params') {
+    // For try builds, we DO NOT want to overwrite parameters, as try builds
+    // offer a superset of PR/nightly parameters, and the extra ones will be
+    // lost when setRobotestParameters() is called -- 2020-04 walt
+    echo "Jenkins Job Parameters:"
       for (param in params) { echo "${param}" }
-      if (env.KEEP_PARAMETERS == 'true') {
-        echo "KEEP_PARAMETERS detected. Ignoring Jenkins job parameters from Jenkinsfile."
+    if (env.KEEP_PARAMETERS == 'true') {
+      echo "KEEP_PARAMETERS detected. Ignoring Jenkins job parameters from Jenkinsfile."
+    } else {
+      echo "Overwriting Jenkins job parameters with parameters from Jenkinsfile."
+      setRobotestParameters()
+      propagateParamsToEnv()
+    }
+  }
+  runRobotest = (env.RUN_ROBOTEST == 'run')
+    stage('build robotest images') {
```

Comment:
I'm not fully sure I understand what you're asking.

I put the build in a separate step to help tracking timing & visualization.  The `run` target will build all images necessary if they're not already complete -- and we could call it directly without this pre-bui.d But I found it useful see whether building images or running robotest was the problem.

Reply:
No, it just looked as if `runRobotest = ...` was starting a new scope where the following `stage` was nested. I just meant it could be on the same indentation level:
```
runRobotest = (env.RUN_ROBOTEST == 'run')
stage('build robotest images') {
...
```

Reply:
Thanks! Fixed in 4720bf4.